### PR TITLE
opcode CODESIZE

### PR DESCRIPTION
**Pull request by roynalnaruto**

### CODE REVIEW

File: src/zkevm_specs/evm/execution_state.py [hunk 1]
```diff
@@ -51,8 +51,8 @@ class ExecutionState(IntEnum):
     CALLDATALOAD = auto()
     CALLDATASIZE = auto()
     CALLDATACOPY = auto()
-    CODESIZE = auto()
     CODECOPY = auto()
+    CODESIZE = auto()
```
[the remark this answers is not in the document]
Comment (roynalnaruto) [reply src/zkevm_specs/evm/execution_state.py:55]: This diff probably slipped in without me noticing. I'll revert this.

File: src/zkevm_specs/evm/execution/codesize.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
+from ...util import N_BYTES_U64
+from ..instruction import Instruction, Transition
+from ..opcode import Opcode
+
+
+def codesize(instruction: Instruction):
+    opcode = instruction.opcode_lookup(True)
+    instruction.constrain_equal(opcode, Opcode.CODESIZE)
+
+    code_size = instruction.bytecode_length(instruction.curr.code_source)
```
[the remark this answers is not in the document]
Comment (roynalnaruto) [reply src/zkevm_specs/evm/execution/codesize.py:10]: @ed255 I've written down my opinion for #73 . I'll wait for others to review/respond and once we agree upon a solution, I can modify this PR.

From implementation perspective, specifically, I am in favour of changing `code_source` to `code_hash`. Hence in the case of contract creation transactions, we populate the bytecode table with the tx calldata (root case) or the caller memory (internal case) and hash the bytecode to update `code_hash` in the current call.

What do you think?

[the remark this answers is not in the document]
Comment (ed255) [reply src/zkevm_specs/evm/execution/codesize.py:10]: > From implementation perspective, specifically, I am in favour of changing `code_source` to `code_hash`. Hence in the case of contract creation transactions, we populate the bytecode table with the tx calldata (root case) or the caller memory (internal case) and hash the bytecode to update `code_hash` in the current call.
> 
> What do you think?

I've also replied to #73 :)  I'm also in favor of replacing `code_source` to `code_hash` and populating the bytecode table the way you mention!

**Comment by roynalnaruto** [comment]
@ed255 #205 is merged, so we can review this PR again.

At the moment I am working on taking #194 over the line so that the pending case specified in #73 [here](https://github.com/privacy-scaling-explorations/zkevm-specs/issues/73#issuecomment-1125754364) is also covered (i.e. copying the init code to the bytecode circuit)

**Review by ed255 — APPROVED** [review]
LGTM!

**Review by miha-stopar — APPROVED** [review]
LGTM!

**Pull request merged**